Jenkinsfile

```diff
@@ -40,13 +40,50 @@ def allArchitectures = [
 ]
 jobProperties.add(parameters([text(defaultValue: allArchitectures.join('\n'),
         description: 'The architectures (cheribuild suffixes) to build for (one per line)',
-        name: 'architectures')]))
+        name: 'architectures'),
+	text(defaultValue: ["riscv64-hybrid", "riscv64-purecap"].join('\n'),
+        description: 'The architectures (cheribuild suffixes) to build with purecap kernel for (one per line)',
+        name: 'purecap_kernel_architectures')]))
 // Set the default job properties (work around properties() not being additive but replacing)
 setDefaultJobProperties(jobProperties)
 
 jobs = [:]
 
-def buildImageAndRunTests(params, String suffix) {
+def selectedPurecapKernelArchitectures = params.purecap_kernel_architectures.split('\n')
+
+def makeTestStep(params, String suffix, testExtraArgs, extraArgs) {
+    return {
+        node {
+            sh label: "Run tests in QEMU", script: """
+    rm -rf test-results && mkdir -p test-results/${suffix}
+    # The test script returns 2 if the tests step is unstable, any other non-zero exit code is a fatal error
+    exit_code=0
+    ./cheribuild/jenkins-cheri-build.py --test run-${suffix} '--test-extra-args=${testExtraArgs.join(" ")}' ${params.extraArgs} ${extraArgs.join(" ")} --test-ssh-key \$WORKSPACE/id_ed25519.pub || exit_code=\$?
+    if [ \${exit_code} -eq 2 ]; then
+        echo "Test script encountered a non-fatal error - probably some of the tests failed."
+    elif [ \${exit_code} -ne 0 ]; then
+        echo "Test script got fatal error: exit code \${exit_code}"
+        exit \${exit_code}
+    fi
+    find test-results
+    """
+            def summary = junitReturnCurrentSummary allowEmptyResults: false, keepLongStdio: true, testResults: "test-results/${suffix}/*.xml"
+            def testResultMessage = "Test summary: ${summary.totalCount}, Failures: ${summary.failCount}, Skipped: ${summary.skipCount}, Passed: ${summary.passCount}"
+            echo("${suffix}: ${testResultMessage}")
+            if (summary.passCount == 0 || summary.totalCount == 0) {
+                params.statusFailure("No tests successful? ${testResultMessage}")
+            } else if (summary.failCount != 0) {
+                // Note: Junit set should have set stage/build status to unstable already, but we still need to set
+                // the per-configuration status, since Jenkins doesn't have a build result for each parallel branch.
+                params.statusUnstable("Unstable test results: ${testResultMessage}")
+                // If there were test failures, we archive the JUnitXML file to simplify debugging
+                archiveArtifacts allowEmptyArchive: true, artifacts: "test-results/${suffix}/*.xml", onlyIfSuccessful: false
```

Comment:
These are going to clash if both the hybrid and purecap kernels fail tests in the same cheribsdtest-foo binary. We probably want a way to add a suffix to (or manually specify) the directory so we can have -purecap-kernel tacked on the end.

Reply:
Yeah, I already had a local patch for this, adding the "-purecap-kernel" suffix.